### PR TITLE
Fix Traefik reload if Consul Catalog tags change

### LINKED ISSUE #2370: Traefik does not reload changes in Consul Catalog tags
*bug*

### What did you do?
Traefik 1.4.2 does not seem to gracefully reload changes to Consul Catalog tags (which again drives the frontend/backend configuration of Traefik. This can be observed by:

1. Configure a service with a setting such as circuit breaker:
```json
        "tags": [
            "traefik.frontend.rule=Host:stuff.stuff.com",
            "traefik.frontend.entryPoints=http",
            "traefik.enable=true",
            "traefik.backend.circuitbreaker=NetworkErrorRatio() > 0.5"
        ]
```

If all is set up correctly, Traefik should populate frontend/backend rules for the service.

In order to see the issue, change a setting in the tags, for example to:
```json
        "tags": [
            "traefik.frontend.rule=Host:stuff.stuff.com",
            "traefik.frontend.entryPoints=http",
            "traefik.enable=true",
            "traefik.backend.circuitbreaker=ResponseCodeRatio(500, 600, 0, 600) > 0.5"
        ]
```

On the node offering the service, issue a consul reload in order to pick up the changes.
If using the Consul Web UI, the ui should now show the updated tags.

However, Traefik still uses the config from the first tag array, and has not updated. When testing the same with Traefik 1.3.8 it correctly updates itself, but on 1.4.2 it does not.

### What did you expect to see?
Traefik should update its config based on the updated Consul tags

### What did you see instead?
Traefik is using outdated Consul tags. Only if Traefik is restarted, it will grab the new tag set and configure itself correctly.

### Output of `traefik version`: (_What version of Traefik are you using?_)
```
Version:      v1.4.2
Codename:     roquefort
Go version:   go1.9.2
Built:        2017-11-02_01:47:46PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
traefikLogsFile = "/var/log/traefik/traefik.log"
checkNewVersion = false
ProvidersThrottleDuration = "5s"
logLevel = "DEBUG"


[retry]

[forwardingTimeouts]
dialTimeout = "5s"
responseHeaderTimeout = "10s"

defaultEntryPoints = ["http"]
[entryPoints]
    [entryPoints.http]
    address = ":80"

[web]
address = ":8080"

[consulCatalog]
constraints = ["tag==environment:uat","tag==lb_id:external-api01"]
watch = true
endpoint = "127.0.0.1:8500"
exposedByDefault = false
```

### PR DESCRIPTION
**Pull request by mmatur**
### What does this PR do?

This PR fix Træfik's reload if Consul Catalog tags change

### More

- [x] Added/updated tests

Fixes #2370

@trondhindenes Could you please try with my experimental image `mmatur/traefik:fix-consul-tag-change`

### CODE REVIEW

File: provider/consul/consul_catalog.go [hunk 1]
```diff
@@ -77,30 +77,41 @@ func (a nodeSorter) Less(i int, j int) bool {
...
 
-func getChangedServiceKeys(currState map[string]Service, prevState map[string]Service) ([]string, []string) {
-	currKeySet := fun.Set(fun.Keys(currState).([]string)).(map[string]bool)
-	prevKeySet := fun.Set(fun.Keys(prevState).([]string)).(map[string]bool)
+func hasChanged(current map[string]Service, previous map[string]Service) bool {
+	addedServiceKeys, removedServiceKeys := getChangedServiceKeys(current, previous)
+	if len(removedServiceKeys) > 0 || len(addedServiceKeys) > 0 {
+		return true
+	}
+	return hasNodeOrTagsChanged(current, previous)
```
Comment (nmengin) [inline provider/consul/consul_catalog.go]: Can be merge in `len(removedServiceKeys) > 0 || len(addedServiceKeys) > 0 || hasNodeOrTagsChanged(current, previous)`?

File: provider/consul/consul_catalog.go [hunk 2]
```diff
@@ -242,19 +253,17 @@ func (p *CatalogProvider) watchCatalogServices(stopCh <-chan struct{}, watchCh c
...
 				// It is possible that there was an idempotent write that does not affect the result of the query.
 				// Thus it is required to do extra check for changes...
-				addedServiceKeys, removedServiceKeys := getChangedServiceKeys(current, flashback)
+				changed := hasChanged(current, flashback)
 
-				addedServiceNodeKeys, removedServiceNodeKeys := getChangedServiceNodeKeys(current, flashback)
-
-				if len(removedServiceKeys) > 0 || len(removedServiceNodeKeys) > 0 || len(addedServiceKeys) > 0 || len(addedServiceNodeKeys) > 0 {
-					log.WithField("MissingServices", removedServiceKeys).WithField("DiscoveredServices", addedServiceKeys).Debug("Catalog Services change detected.")
+				if changed {
```
Comment (nmengin) [inline provider/consul/consul_catalog.go]: Maybe use directly the method in the `if` statement?

`if hasChanged(current, flashback) {`

**Comment by trondhindenes** [comment]
awesome @mmatur, thanks for looking at this so quickly!

**Comment by mmatur** [comment]
@trondhindenes you can easily extract binary from the container
```shell
docker run -tid --name=traefik-experimental mmatur/traefik:fix-consul-tag-change
docker cp traefik-experimental:/traefik .
```

**Comment by trondhindenes** [comment]
thanks!

**Comment by trondhindenes** [comment]
Hm, the web ui is giving me a 404 with this build. Is that intended?
```
[web]
address = ":8080"
```
It redirects from root to `/dashboard` and then I get a 404 from there.

**Comment by trondhindenes** [comment]
Looking at live traefik logs while doing a deploy now I notice that Traefik is still not reloading its config when  the service is flagged as "in maintenance using the `consul maint` command. Same as before: The Consul web ui is updated, clearly showing the correct state, but Traefik is not honoring maintenance mode.

**Comment by mmatur** [comment]
For the webui it is because I have not build the webui when I have build my docker image. 
Traefik does not watch the maintenance mode. Could you please create an issue to request a feature.

**Comment by trondhindenes** [comment]
I'm also still seeing the behavior of Traefik not responding to changes in the service in general. I just tried rebooting a backend. Traefik updated its config when the node left, but not when it came back online (consul web ui shows that the consul db is up to date, and running `consul watch -service=myservicename -type=service` locally on the traefik node shows everything correct after the backend node has come back online.

**Comment by trondhindenes** [comment]
"Traefik does not watch the maintenance mode" - okay that explains a good deal. That should probably go into documentation in a very clear way, as it's a core feature of Consul.

**Comment by trondhindenes** [comment]
Still, it doesnt explain all the behavior I'm seeing. It should watch for `node leave` events in Consul right?

**Comment by mmatur** [comment]
Could you give me your last configuration of traefik and tags for each services. I will investigate. And I will try to give you a new traefik docker image with webui.

**Comment by trondhindenes** [comment]
Sure, 
btw don't worry about the web ui. The rest interface in the image you gave me is working, and that's more than enough to see what Traefik is doing.

traefik.toml:
```
traefikLogsFile = "/var/log/traefik/traefik.log"
checkNewVersion = false
ProvidersThrottleDuration = "5s"
logLevel = "DEBUG"

defaultEntryPoints = ["http"]
[entryPoints]
    [entryPoints.http]
    address = ":80"

[web]
address = ":8080"

[web.metrics.datadog]
address = "localhost:8125"
pushinterval = "10s"

[consulCatalog]
constraints = ["tag==environment:uat","tag==lb_id:external-api01"]
watch = true
endpoint = "127.0.0.1:8500"
exposedByDefault = false
```

Output from curl `http://localhost:8500/v1/catalog/service/services-external-api-uat-main-playapi | python -m json.tool`

```
[
    {
        "Address": "10.245.12.4",
        "CreateIndex": 4978547,
        "ID": "f54dd745-d6d2-4695-b67a-9e89e20b43ce",
        "ModifyIndex": 4978547,
        "Node": "wuca-0202003",
        "NodeMeta": {},
        "ServiceAddress": "",
        "ServiceEnableTagOverride": false,
        "ServiceID": "services-external-api-uat-main-playapi",
        "ServiceName": "services-external-api-uat-main-playapi",
        "ServicePort": 80,
        "ServiceTags": [
            "environment:uat",
            "traefik.frontend.rule=Host:play-api.uat.domain.com;PathPrefixStrip: /play/1",
            "traefik.frontend.entryPoints=http",
            "traefik.enable=true",
            "traefik.tags=environment:uat,lb_id:external-api01",
            "traefik.backend.weight=30"
        ],
        "TaggedAddresses": {
            "lan": "10.245.12.4",
            "wan": "10.245.12.4"
        }
    },
    {
        "Address": "10.245.12.26",
        "CreateIndex": 4932296,
        "ID": "0ea3fb7e-1a81-4539-b457-a9d79a121cf9",
        "ModifyIndex": 4932296,
        "Node": "wum-0202020",
        "NodeMeta": {},
        "ServiceAddress": "",
        "ServiceEnableTagOverride": false,
        "ServiceID": "services-external-api-uat-main-playapi",
        "ServiceName": "services-external-api-uat-main-playapi",
        "ServicePort": 80,
        "ServiceTags": [
            "environment:uat",
            "traefik.frontend.rule=Host:play-api.uat.domain.com;PathPrefixStrip: /play/1",
            "traefik.frontend.entryPoints=http",
            "traefik.enable=true",
            "traefik.tags=environment:uat,lb_id:external-api01",
            "traefik.backend.weight=90"
        ],
        "TaggedAddresses": {
            "lan": "10.245.12.26",
            "wan": "10.245.12.26"
        }
    }
]

```

**Comment by mmatur** [comment]
Thanks for your configuration. Could you please send the the full debug log of your traefik instance?

**Comment by trondhindenes** [comment]
Which version do you want the debug log from? The one from this custom build? Can do. But in general it doesn't say much, you can just see that it does _not_ reload configuration so it just sits there. But I'm happy to send it over if you confirm the version.

**Comment by mmatur** [comment]
@trondhindenes Yes the log frmo the custom build. To have more verbose logs please add `debug = true` in your traefik toml configuration

**Comment by trondhindenes** [comment]
Sure, will do - I'll send those over tomorrow morning. I saw this getting merge and the related issue was closed but just FYI: I'm still having issues with Consul not reloading config on changes. Should we track this in another issue? Tracking stuff in a merged PR seems a bit sub-optimal.

**Comment by mmatur** [comment]
@trondhindenes Could you please create a new issue with last configuration and traefik debug log

**Comment by zaargy** [comment]
Ah well I am still seeing this issue in 1.5.4 release as well.

**Comment by mmatur** [comment]
Hi @zaargy,

Could you have a try with last Træfik version [v1.6.3](https://github.com/containous/traefik/releases/tag/v1.6.3). 
And if you encountered an issue could you please open new [GitHub Issue](https://github.com/containous/traefik/issues/new/choose).

**Pull request merged**